src/content/blog/2023-09-26-preparing-custom-rules-eslint-v9.md

```diff
@@ -0,0 +1,245 @@
+
+To recreate this data, you'll need to track the traversal state manually, which can be accomplished with the following code:
+
+```js
+module.exports = {
+    meta: {
+        // ...
+    },
+    create(context) {
+
+        // tracks the code path we are currently in
+        let currentCodePath;
+
+        // tracks the segments we've traversed in the current code path
+        let currentSegments;
+
+        // tracks all current segments for all open paths
+        const allCurrentSegments = [];
+
+        return {
+
+            onCodePathStart(codePath) {
+                currentCodePath = codePath;
+                allCurrentSegments.push(currentSegments);
+                currentSegments = new Set();
+            },
+
+            onCodePathEnd(codePath) {
+                currentCodePath = codePath.upper;
+                currentSegments = allCurrentSegments.pop();
+            },
+
+            onCodePathSegmentStart(segment) {
+                currentSegments.add(segment);
+            },
+
+            onCodePathSegmentEnd(segment) {
+                currentSegments.delete(segment);
+            },
+
+            onUnreachableCodePathSegmentStart(segment) {
+                currentSegments.add(segment);
+            },
+
+            onUnreachableCodePathSegmentEnd(segment) {
+                currentSegments.delete(segment);
+            }
+        };
+
+    }
+};
+```
+
+We have already made this change in all of the ESLint core rules to validate that the approach works as expected.
+
+## `context` methods becoming properties
+
+As we look towards the API we'd like rules for other languages to have, we decided to convert some methods on `context` to properties. The methods in the following table just returned some data that didn't change, so there was no reason they couldn't be properties instead.
+
+|**Deprecated on `context`**|**Property on `context`**|
+|-----------------------|--------------------------|
+|`context.getSourceCode()`|`context.sourceCode`|
+|`context.getFilename()`|`context.filename`|
+|`context.getPhysicalFilename()`|`context.physicalfilename`|
+|`context.getCwd()`|`context.cwd`|
+
+We are deprecating the methods in favor of the properties (added in v8.40.0). These methods will be removed in v10.0.0 (not v9.0.0) as they are not blocking language plugins work. Here's an example that ensures the correct value is used:
+
+```js
+module.exports = {
+    create(context) {
+
+        const sourceCode = context.sourceCode ?? context.getSourceCode();
+        const cwd = context.cwd ?? context.getCwd();
+        const filename = context.filename ?? context.getFilename();
+        const physicalFilename = context.physicalFilename ?? context.getPhysicalFilename();
+
+        return {
+            Program(node) {
+                // do something
+            }
+        }
+    }
+};
+```
+
+## `context` properties being removed
```

Comment:
In https://github.com/eslint/eslint/issues/13481, removing `context.parserPath` and `context.parserOptions` is in Phase 5: Remove `eslintrc`, which would be in eslint v10, not v9?

Reply:
Good call